torchx/schedulers/ray_scheduler.py

```diff
@@ -0,0 +1,299 @@
+        num_gpus:
+            The number of GPUs to allocate.
+        memory_size:
+            The amount of memory, in MBs, to allocate.
+    """
+
+    name: str
+    command: str
+    env: Dict[str, str] = field(default_factory=dict)
+    num_replicas: int = 1
+    num_cpus: int = -1
+    num_gpus: int = -1
+    memory_size: int = -1
+    # TODO: max_retries, retry_policy
+
+
+@dataclass
+class RayJob:
+    """Represents a job that should be run on a Ray cluster.
+
+    Attributes:
+        app_id:
+            The unique ID of the application (a.k.a. job).
+        cluster_config_file:
+            The Ray cluster configuration file.
+        cluster_name:
+            The cluster name to use.
+        start_cluster:
+            A boolean value indicating whether to start the cluster if needed.
+        copy_script:
+            A boolean value indicating whether to copy the script files to the
+            cluster.
+        copy_script_dir:
+            A boolean value indicating whether to copy the directories
+            containing the scripts to the cluster.
+        scripts:
+            The set of scripts to copy to the cluster.
+        actors:
+            The Ray actors which represent the job to be run. This attribute is
+            dumped to a JSON file and copied to the cluster where `ray_main.py`
+            uses it to initiate the job.
+        verbose:
+            A boolean value indicating whether to enable verbose output.
+    """
+
+    app_id: str
+    cluster_config_file: str
+    cluster_name: Optional[str] = None
+    start_cluster: bool = False
+    copy_scripts: bool = False
+    copy_script_dirs: bool = False
+    scripts: Set[str] = field(default_factory=set)
+    actors: List[RayActor] = field(default_factory=list)
+    verbose: bool = False
+
+
+class RayScheduler(Scheduler):
+    def __init__(self, session_name: str) -> None:
+        super().__init__("ray", session_name)
+
+    def run_opts(self) -> runopts:
+        opts = runopts()
+        opts.add(
+            "cluster_config_file",
+            type_=str,
+            required=True,
+            help="Use CLUSTER_CONFIG_FILE to access or create the Ray cluster.",
+        )
+        opts.add(
+            "cluster_name",
+            type_=str,
+            help="Override the configured cluster name.",
+        )
+        opts.add(
+            "start_cluster",
+            type_=bool,
+            default=False,
+            help="Start the cluster if needed.",
+        )
+        opts.add(
+            "copy_scripts",
+            type_=bool,
+            default=False,
+            help="Copy the script(s) to the cluster. Supported only for PyTorch DDP jobs.",
+        )
+        opts.add(
+            "copy_script_dirs",
+            type_=bool,
+            default=False,
+            help="Copy the directories containing the scripts to the cluster. "
+            "Supported only for PyTorch DDP jobs.",
```

Comment:
@aivanou this seems similar to the patch behavior you've been looking to add internally